clevis.go

```diff
@@ -101,23 +98,38 @@ func (p Pin) ToConfig() (Config, error) {
 
 // Decrypt decrypts a clevis bound message. The message format can be either compact or JSON.
 func Decrypt(data []byte) ([]byte, error) {
-	msg, p, err := Parse(data)
+	return jwe.Decrypt(data, "", nil, jwe.WithPostParser(clevisParser{}))
+}
+
+type clevisParser struct{}
+
+// Do implements a jwe.PostParser interface which must use the excracted pin config to set up the decryprion algorithm and key
+func (clevisParser) Do(ctx jwe.DecryptCtx) error {
+	p, err := pinFromMsg(ctx.Message())
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	switch p.Pin {
 	case "tang":
-		return p.Tang.decrypt(msg)
+		err = p.Tang.prepareDecryptionCtx(ctx)
```

Comment:
The function `prepareDecryptionCtx()` computes the key and sets to the decryptcontext. I prefer to split these 2 operations. `prepareDecryptionCtx` receives `jwx.Message` and returns `algorithm, key, error`. And verifying/setting algo/key is done after the switch statement.

`prepareDecryptionCtx` could be renamed to something like `computeKey`.

It also makes sense to move the whole `switch` statement to a separate function under Pin struct, e.g. `Pin.computeKey()`.

Reply:
I also work on making `tang` key computation more consistent with the rest of the code. With this commits https://github.com/anatol/clevis.go/tree/drop-frankenstein-ecmr all pin implementations for the `computeKey()` will return `jwa.DIRECT, []byte` where the second parameter is the key. In this case `computeKey()` signature could be further simplified to `[]byte,error`.

Reply:
I was thinking along the same lines.  I like that approach for the real implementation... I was just trying to get something to work workout putting a lot of thought into it :)

But yes, if each specific pin implemented a `recoverKey()  ([]byte, error)` that would be much cleaner!

Reply:
In this case I suggest to go with the tang refactoring https://github.com/anatol/clevis.go/pull/9 first so your code can be simplified a bit.

Reply:
Agreed! I'll gladly clean this up once that's in.